numpy/core/src/umath/ufunc_object.c

```diff
@@ -4054,6 +4054,7 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc,
         }
         Py_INCREF(dtype->singleton);
         otype = dtype->singleton;
+        Py_DECREF(dtype);
```

Comment:
A bit of explanation would be good here. I assume that dtype already holds a reference to the singleton member and cleans that up when dtype is destroyed. True? Might read better  like 
```
       otype = dtype->singleton;
        Py_INCREF(otype);
        Py_DECREF(dtype);
```

Reply:
Yeah, that is better.  This is a bit weird `singleton` is guaranteed to exist and `dtype` can never be destroyed because this is limited to "old-style" DTypes.

This code is simply not yet new-style user `DType` ready.  The whole dance will vanish (or rather: move into legacy-only fallback branches) with the ufunc refactor.  Until then, this is a bit awkward because I try to introduce the new logic step-by-step.